### PR TITLE
build: enable libc++ hardening in debug builds

### PR DESCRIPTION
**Pull request by vasild**
When compiling with libc++ in debug mode, then enable full libc++ hardening.

Inspired by https://github.com/bitcoin/bitcoin/issues/31272#issuecomment-2518700939

---

Note that `libstdc++`'s `_GLIBCXX_ASSERTIONS` (aka light debug mode) is [enabled by default](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_macros.html) when compiling without optimizations), that is in our debug builds. But see https://github.com/bitcoin/bitcoin/pull/31424#issuecomment-2732100998.

---

Further considerations:

* Consider enabling `libstdc++`s `_GLIBCXX_ASSERTIONS` (aka light debug mode) also in non-debug builds. Should assess [performance impact](https://github.com/bitcoin/bitcoin/pull/31424#discussion_r1910658664).

* Consider enabling `libc++`s `_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_FAST` in non-debug builds. Should assess peformance impact as well.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -459,6 +459,12 @@ try_append_cxx_flags("-fstack-reuse=none" TARGET core_interface)
 if(ENABLE_HARDENING)
   add_library(hardening_interface INTERFACE)
   target_link_libraries(core_interface INTERFACE hardening_interface)
+
+  target_compile_options(hardening_interface INTERFACE
+    -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG
+    -D_GLIBCXX_ASSERTIONS
+  )
```
Comment (vasild) [inline CMakeLists.txt:466]: Should this be set to `_LIBCPP_HARDENING_MODE_FAST` or `_LIBCPP_HARDENING_MODE_DEBUG` based on the build type?

File: CMakeLists.txt [hunk 2]
```diff
@@ -459,6 +459,12 @@ try_append_cxx_flags("-fstack-reuse=none" TARGET core_interface)
...
+
```
Comment (vasild) [inline CMakeLists.txt:467]: How is this related to the following?

https://github.com/bitcoin/bitcoin/blob/e8cc790fe2a223cfaa6dedfba88a6dc4f336f1d9/depends/hosts/linux.mk#L16-L20

**Review by vasild — COMMENTED** [review]
cc @hebasto

File: CMakeLists.txt (hunk 2 above)
Comment (dergoegge) [reply CMakeLists.txt:467]: These are flags set for `DEBUG=1` when building depends. Iirc `-D_GLIBCXX_DEBUG` includes `-D_GLIBCXX_ASSERTIONS` as well as other slower checks.

File: CMakeLists.txt [hunk 3]
```diff
@@ -459,6 +459,12 @@ try_append_cxx_flags("-fstack-reuse=none" TARGET core_interface)
 if(ENABLE_HARDENING)
   add_library(hardening_interface INTERFACE)
   target_link_libraries(core_interface INTERFACE hardening_interface)
+
+  target_compile_options(hardening_interface INTERFACE
```
Comment (dergoegge) [inline CMakeLists.txt:463]: Maybe we should add this to depends instead?

File: CMakeLists.txt (hunk 1 above)
Comment (vasild) [reply CMakeLists.txt]: I think so, "yes". Done.

File: CMakeLists.txt (hunk 2 above)
Comment (vasild) [reply CMakeLists.txt]: > How is this related to the following?

I understand that the flags in `bitcoin/depends/hosts/linux.mk` are (obviously) only used when building depends and only on Linux. In comparison, if added to `CMakeLists.txt` under `ENABLE_HARDENING` then the flags will be used for depends and for non-depends builds on all platforms that use libc++ or libstdc++, not only Linux. Thus I dropped the flags from `bitcoin/depends/hosts/linux.mk` and added (a more extensive) hardening to `CMakeLists.txt`.

Yes, `-D_GLIBCXX_DEBUG` includes `-D_GLIBCXX_ASSERTIONS`.

> How do they interact?

I guess it is harmless to specify both `_GLIBCXX_DEBUG*` and `_LIBCPP_HARDENING_MODE` because libstdc++ will honor the first and libc++ the second. 

> I don't think it is useful to possibly silently downgrade DEBUG into FAST.

Indeed that would be bad. If `_LIBCPP_HARDENING_MODE` is provided more than once on the command line, then a warning will be raised: "macro is redefined". E.g. `-D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_FAST` will produce a warning.

> Also, it can be considered to enable the ABI checks

Done.

File: CMakeLists.txt (hunk 3 above)
Comment (vasild) [reply CMakeLists.txt]: By default hardening is enabled in the CMake config and the depends build will benefit from that. Only if depends build uses `make NO_HARDEN=1` then it will disable the hardening in CMake.

File: CMakeLists.txt [hunk 4]
```diff
@@ -512,6 +512,41 @@ if(ENABLE_HARDENING)
...
+    if(HAVE_LIBCPP)
+      # https://libcxx.llvm.org/Hardening.html
+      if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+        target_compile_options(hardening_interface INTERFACE
+          -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS_IN_STRING
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS_IN_VECTOR
+          -D_LIBCPP_ABI_BOUNDED_UNIQUE_PTR
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS_IN_STD_ARRAY
```
[the remark this answers is not in the document]
Comment (vasild) [reply CMakeLists.txt:525]: Not manually, but I assume CI will fail if that happens. If it fails, then I will remove it.

File: CMakeLists.txt (hunk 2 above)
Comment (fanquake) [reply CMakeLists.txt]: > I don't think it is useful to possibly silently downgrade DEBUG into FAST.

> If _LIBCPP_HARDENING_MODE is provided more than once on the command line, then a warning will be raised

Just printing a warning will more than likely still result in silent downgrading.

Comment (vasild) [reply CMakeLists.txt]: `-Werror` would flip that warning into an error. But anyway, in the latest incarnation of this PR `_LIBCPP_HARDENING_MODE` is set from one place. So there is no way that one part of the build system setting it to one value and another part of the build system setting it to another value, overriding the first one.

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vasild) [reply CMakeLists.txt:525]: Do you think it would be better to omit those `_LIBCPP_ABI_BOUNDED*`?

**Review by maflcko — CHANGES_REQUESTED** [review]
There will need to be a workaround for gcc12-14, see https://github.com/bitcoin/bitcoin/issues/31436

Edit: Fixed in https://github.com/bitcoin/bitcoin/pull/31493

File: CMakeLists.txt [hunk 5]
```diff
@@ -512,6 +512,38 @@ if(ENABLE_HARDENING)
     if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
       try_append_linker_flag("-Wl,-fixup_chains" TARGET hardening_interface)
     endif()
+
+    if(HAVE_LIBCPP)
+      # https://libcxx.llvm.org/Hardening.html
+      if(CMAKE_BUILD_TYPE STREQUAL "Debug")
```
Comment (theuni) [inline CMakeLists.txt:518]: See [here](https://github.com/bitcoin/bitcoin/pull/31458#issuecomment-2541429701) for a more robust approach to the if(debug) test.

File: CMakeLists.txt (hunk 2 above)
Comment (vasild) [reply CMakeLists.txt]: The libc hardening is now removed from `depends/hosts/linux.mk` in favor of more universal hardening from `CMakeLists.txt` (which is enabled when building via depends).

File: CMakeLists.txt (hunk 5 above)
Comment (vasild) [reply CMakeLists.txt:518]: Yeah, I considered `core_interface_debug`, but I couldn't fit it into this pattern:

```
if debug then
    add compile-definitions-A to hardening_interface
else
    add compile-definitions-B to hardening_interface
```

because

1. the surrounding code adds stuff to `hardening_interface` and I wanted to do the same, not add to `core_interface_debug` and

2. if I add to `core_interface_debug` then in the non-debug case (the `else` branch), where should I add `compile-definitions-B` to? `core_interface` will be used for both release and debug. I think we don't have a symmetric `core_interface_release` which will only be used in release, like `core_interface_debug` is used only in debug.

File: CMakeLists.txt [hunk 6]
```diff
@@ -512,6 +512,38 @@ if(ENABLE_HARDENING)
...
+        target_compile_definitions(hardening_interface INTERFACE
+          _LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG
+          _LIBCPP_ABI_BOUNDED_ITERATORS
+          _LIBCPP_ABI_BOUNDED_ITERATORS_IN_STRING
+          _LIBCPP_ABI_BOUNDED_ITERATORS_IN_VECTOR
+          _LIBCPP_ABI_BOUNDED_UNIQUE_PTR
+          _LIBCPP_ABI_BOUNDED_ITERATORS_IN_STD_ARRAY
+        )
+      else()
+        target_compile_definitions(hardening_interface INTERFACE _LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_FAST)
```
[the remark this answers is not in the document]
Comment (vasild) [reply CMakeLists.txt:528]: Seems like you copied your comment from https://github.com/bitcoin/bitcoin/pull/31424#discussion_r1871047821 to a new discussion thread here. I think that is confusing and will make it difficult to follow the conversation flow. I have already replied to your comment in the [original thread](https://github.com/bitcoin/bitcoin/pull/31424#discussion_r1875891496).

I am marking this as resolved. If you think the original thread was wrongly marked as resolved I think it would be better to comment there and I will reopen it (or feel free to reopen it yourself if github allows that).

File: CMakeLists.txt (hunk 2 above)
Comment (vasild) [reply CMakeLists.txt]: How would a vendor enable `_LIBCPP_HARDENING_MODE`? Some way other than `cmake -DENABLE_HARDENING=ON`? If they supply flags somehow externally, then the same argument can be made for any flag - our build system may override a flag set externally (e.g. `-O1` vs `-O2` or whatever).

Thinking about this, the proper way to supply flags externally is to use `cmake -DAPPEND_CPPFLAGS="-DFOO=1"`.

I think this is safe:

* The proper way to enable `_LIBCPP_HARDENING_MODE` and other hardening options depending on compiler and environment is via `cmake -DENABLE_HARDENING=ON`.
  * If somebody uses `cmake -DENABLE_HARDENING=ON` _and_ supplies `_LIBCPP_HARDENING_MODE` externally, then they are shooting themselves in the foot and will get compilation warning or error depending on `-Werror` about redefined macro. The warnings will be all over the place, for every invocation of the compiler for every file.
* It would be possible to fiddle manually with this by using `cmake -DENABLE_HARDENING=OFF -DAPPEND_CPPFLAGS="-D_LIBCPP_HARDENING_MODE=XYZ"`

Comment (fanquake) [reply CMakeLists.txt]: > Thinking about this, the proper way to supply flags externally is to use cmake -DAPPEND_CPPFLAGS="-DFOO=1".

No? The append_* flags are a hack we had to add to bypass various other CMake behaviours and they are not intended for general use (hence "special builds").

Comment (vasild) [reply CMakeLists.txt]: > I meant a libc++ vendor

Oh, sorry for misunderstanding.

So, according to https://libcxx.llvm.org/Hardening.html#notes-for-users the libc++ vendor can set a default value and users that wish to use a different value can do that by passing `-D_LIBCPP_HARDENING_MODE=...` to the compiler which is exactly what Bitcoin Core will do. Sounds good?

Comment (vasild) [reply CMakeLists.txt]: Using an example would help. Are you worried about the following?

A libc++ vendor, lets say "Ubuntu 123.45 LTS", compiles the libc++ they ship with `LIBCXX_HARDENING_MODE=debug` (see https://libcxx.llvm.org/Hardening.html#notes-for-vendors). Then a user of "Ubuntu 123.45 LTS" compiles Bitcoin Core with `cmake -DCMAKE_BUILD_TYPE=Release`. This results in Bitcoin Core being compiled with `fast` hardening mode.

The above is how it should be - the OS default is supposed to be overridable and in Bitcoin Core `-DCMAKE_BUILD_TYPE=Release` is intended to use the `fast` mode.

There is also this note in the docs:

> Since the static and shared library components of libc++ are built by the vendor, setting this macro will have no impact on the hardening mode for the pre-built components. Most libc++ code is header-based, so a user-provided value for _LIBCPP_HARDENING_MODE will be mostly respected.

I think it is no-brainer that an application (e.g. Bitcoin Core) should pass `-D_LIBCPP_HARDENING_MODE=...` to the compiler and that is the correct and the best thing an app can do and this PR does that.

Comment (vasild) [reply CMakeLists.txt]: Extended the PR description, see the <sup>1</sup> specifically. I wouldn't call it a "downgrade" - it is less checks, but an "upgrade" from performance point of view. All these settings provide varying tradeoffs between performance and amount of checks. Downgrade for one is upgrade for the other.

This is analogous to the `-O` flag. Assuming the default is `-O1` (if no `-O...` is provided). If the user explicitly provides `-O2` that will use `-O2` instead of the default `-O1`. I wouldn't call this (silent) "downgrade" or "upgrade", but rather "intentional override of the default".

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vasild) [reply CMakeLists.txt]: I removed the `_LIBCPP_ABI_BOUNDED*` options from CMake when `ENABLE_HARDENING=ON` because they caused linking failures for the multiprocess task because libmultiprocess was compiled without them, i.e. ABI incompatibility issues. So, don't enable them on users when `ENABLE_HARDENING=ON`.

> It would be good to have them in a CI task

Right, I changed a few CI tasks to use them, enabling them manually, outside of what CMake does by default, see the second commit and the PR description.

[the remark this answers is not in the document]
Comment (vasild) [reply CMakeLists.txt]: No, wait, it was the tsan task, not the multiprocess task! Edited the above comment of mine. Anyway, it does not matter that much - it is an ABI incompatibility, so the conclusion is the same: don't enable `_LIBCPP_ABI_BOUNDED*` on users when `ENABLE_HARDENING=ON`.

> Why would the other tasks not break the ABI?

I am not sure exactly, but it has probably something to do with the way libc++ is compiled or `_LIBCPP_HARDENING_MODE_FAST` not playing well with `_LIBCPP_ABI_BOUNDED*`. Here is the exact failure: https://api.cirrus-ci.com/v1/task/5304194619408384/logs/ci.log, see `undefined reference to ... std::__1::basic_string<...>::insert(std::__1::__bounded_iter<...`. This error just does not happen on the other tasks.

> Just because CI is green doesn't mean the change is correct.

If there are ABI incompatibilities, then it will fail to compile with undefined reference because it cannot find `__bounded_iter` or `std::__debug::whatever` in the standard library.

**Review by maflcko — COMMENTED** [review]
concept ack, but it would be good to get the (CI) changes right

**Comment by vasild** [comment]
`7daae2cf96...39a1326488`: drop some changes and move them into a separate PR: https://github.com/bitcoin/bitcoin/pull/31612, suggested by @maflcko

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vasild) [reply CMakeLists.txt]: All `_LIBCPP_ABI_BOUNDED*` additions have been removed from this PR.

File: depends/hosts/linux.mk [hunk 7]
```diff
@@ -13,12 +13,6 @@ linux_release_CXXFLAGS=$(linux_release_CFLAGS)
 linux_debug_CFLAGS=-O1 -g
 linux_debug_CXXFLAGS=$(linux_debug_CFLAGS)
 
-# https://gcc.gnu.org/onlinedocs/libstdc++/manual/debug_mode.html
-linux_debug_CPPFLAGS=-D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC
```
Comment (fanquake) [inline depends/hosts/linux.mk:17]: Not sure. This just seems to be removing an option (that otherwise doesn't conflict with the other changes)?

File: CMakeLists.txt [hunk 8]
```diff
@@ -521,6 +521,33 @@ if(ENABLE_HARDENING)
...
+
+    if(HAVE_LIBSTDCPP)
+      # https://gcc.gnu.org/wiki/LibstdcxxDebugMode
+      # https://gcc.gnu.org/onlinedocs/libstdc++/manual/debug_mode.html
+      #
+      # libstdc++ has _GLIBCXX_DEBUG _GLIBCXX_DEBUG_PEDANTIC but they are not ABI compatible with
+      # normal mode which could cause linkage failures like:
+      #   undefined reference to `mp::Connection::removeSyncCleanup(__gnu_debug::_Safe_iterator...`
+      # when an external library is not compiled with _GLIBCXX_DEBUG _GLIBCXX_DEBUG_PEDANTIC,
+      # so we use _GLIBCXX_ASSERTIONS in both Debug and non-Debug builds.
```
Comment (fanquake) [inline CMakeLists.txt:542]: > so we use _GLIBCXX_ASSERTIONS in both Debug and non-Debug builds.

How did you test the performance impacts of this? For example master:
```bash
./master/src/bench/bench_bitcoin -filter=LinearizeOptimallyExample11
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|      495,996,518.00 |                2.02 |    0.3% |5,431,375,728.00 |1,800,295,241.00 |  3.017 | 551,650,338.00 |    0.2% |      5.44 | `LinearizeOptimallyExample11`
```
vs this PR:
```bash
./build_assertions/src/bench/bench_bitcoin -filter=LinearizeOptimallyExample11
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|      678,578,986.00 |                1.47 |    0.2% |8,178,898,994.00 |2,465,218,869.00 |  3.318 |1,114,065,676.00 |    0.1% |      7.47 | `LinearizeOptimallyExample11`
```

Also see: https://corecheck.dev/bitcoin/bitcoin/pulls/31424.

File: depends/hosts/linux.mk (hunk 7 above)
Comment (vasild) [reply depends/hosts/linux.mk:17]: I removed this because the same options were enabled by CMake's HARDERNING option when in Debug mode, so this became redundant. However, later I changed HARDENING+Debug to only use `_GLIBCXX_ASSERTIONS` instead of `-D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC` because the latter produced: `undefined reference to mp::Connection::removeSyncCleanup(__gnu_debug::_Safe_iterator...`.

Maybe this should stay. I wonder why using `-D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC` from here does not produce an error...

File: CMakeLists.txt (hunk 8 above)
Comment (vasild) [reply CMakeLists.txt:542]: Interesting. Is that using libstdc++ or libc++?

Comment (fanquake) [reply CMakeLists.txt:542]: libstdc++

**Comment by vasild** [comment]
`1ef9f7ca4a...8cf75a4a18`: use generator expressions instead of `if(CMAKE_BUILD_TYPE STREQUAL "Debug")` because they also work for the `Ninja Multi-Config` generator.

File: CMakeLists.txt (hunk 5 above)
Comment (vasild) [reply CMakeLists.txt]: I ended up appending to `hardening_interface` and using the CMake generator expression `$<IF:condition,true_string,false_string>`. Resolved?

**Comment by fanquake** [comment]
What is the status of this? Are you investigating the performance regressions?

File: CMakeLists.txt (hunk 2 above)
Comment (vasild) [reply CMakeLists.txt]: Dropped changes to `depends/hosts/linux.mk`.

File: depends/hosts/linux.mk (hunk 7 above)
Comment (vasild) [reply depends/hosts/linux.mk]: I didn't realize that these flags are used also for the libraries. Dropped these changes.

File: CMakeLists.txt (hunk 8 above)
Comment (vasild) [reply CMakeLists.txt]: Given the above I left non-debug builds alone.

**Comment by vasild** [comment]
`8cf75a4...a3a799c`: reduce the scope of this to only add hardening on libc++ when in debug mode, updated the OP.

**Comment by vasild** [comment]
Excellent! Removed from OP.

**Comment by vasild** [comment]
The documentation that I linked is explicit:

https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_macros.html

> _GLIBCXX_ASSERTIONS
    Defined by default when compiling with no optimization, undefined by default when compiling with optimization. When defined, enables extra error checking in the form of precondition assertions, such as bounds checking in strings and null pointer checks when dereferencing smart pointers.

If you suspect that there is a bug in GCC or the documentation is wrong, you can try the above `random_tests/t` compiled on `master` (this PR does not change behavior when GCC is used):

`-DCMAKE_BUILD_TYPE=Debug` (uses `-O0`):
```
/usr/local/lib/gcc15/include/c++/bits/stl_vector.h:1262: constexpr std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = int; _Alloc = std::allocator<int>; reference = int&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
```

`-DCMAKE_BUILD_TYPE=Release` (uses `-O2`):
```
*** No errors detected
```

**Comment by vasild** [comment]
Can't enable `_GLIBCXX_DEBUG_PEDANTIC` out in the wild (ie for everybody via CMake). But it is already enabled in a controlled environment (the CI): https://github.com/bitcoin/bitcoin/pull/31424#issuecomment-2724631464

**Comment by vasild** [comment]
> libc++ hardening is also enabled in the CI.

Yes. That is not ABI breaking, so it _can_ be enabled out in the wild (ie for everybody via CMake) in addition to the CI.

> I find it extremely hard to review this pull request, because for months, on every push it is changing what hardening setting to apply in what severity to what part of the system.

I changed it based on reviewers' feedback, including yours. Thanks for that! Also reduced the scope and dropped parts of it because of possible performance issues. Right now it is `+8 -0` lines.

> My recommendation would be to just state your goal, so that reviewers can easily decide what is the best way to achieve your goal.

It is stated in the PR title and description.

**Comment by vasild** [comment]
> I'd just find it odd to enable hardening for libc++, but not for GCC in debug builds.

This conversation is going in circles: can't enable full GCC debug like for Clang because the GCC one is ABI breaking: https://github.com/bitcoin/bitcoin/pull/31424#issuecomment-2772791759

> GCC is probably used more widely by devs. I know it may be enabled by default for the GCC std lib in a future release, but it may take some time until devs are updating to that. libc++ will also do the same: https://www.github.com/llvm/llvm-project/issues/122925

This is about the "fast/light" debug which is not in the scope of this PR (anymore). GCC is not in the scope of this PR either. If you want to do GCC and/or "fast" debug, feel free to open another PR, I will review. That was here and was dropped.

> So if upstream is a reason to not do it for gcc, then it could also be skipped for libc++?

GCC is ABI breaking and Clang is not. So, only do Clang in this PR.

> However, you are enabling the strongest debug mode for libc++ here, so I also wonder about the performance impact? I guess performance expectations are non-existing in debug builds, but it would still be good to know, as it would be impossible to disable, and would disallow hacks such as https://www.github.com/bitcoin/bitcoin/pull/32113#issuecomment-2747808769

Can be disabled with:

`cmake ... -DAPPEND_CPPFLAGS="-D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE"`

**Comment by fanquake** [comment]
Note that this will also appear broken in the summary:
```bash
cmake -B build -DCMAKE_BUILD_TYPE=Debug
<snip>
Preprocessor defined macros ........... <snip> RPC_DOC_CHECK ABORT_ON_FAILED_ASSUME $<IF:$<CONFIG:Debug>,_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG,>
```

**Comment by vasild** [comment]
> I guess performance expectations are non-existing in debug builds, but it would still be good to know ...

Agree. Debug-performance of the unit tests seems to be the same as in `master`, or within noise:

master:
```
$ for i in {1..5} ; do ctest --test-dir build-master 2>&1 |grep 'Total Test time' ; done
Total Test time (real) = 407.20 sec
Total Test time (real) = 397.09 sec
Total Test time (real) = 422.87 sec
Total Test time (real) = 410.20 sec
Total Test time (real) = 415.94 sec
```

PR:
```
$ for i in {1..5} ; do ctest --test-dir build-pr 2>&1 |grep 'Total Test time' ; done
Total Test time (real) = 403.39 sec
Total Test time (real) = 403.48 sec
Total Test time (real) = 404.27 sec
Total Test time (real) = 394.76 sec
Total Test time (real) = 407.62 sec
```

**Comment by fanquake** [comment]
`ENABLE_HARDENING` and the `hardening_interface` have been removed.

**Comment by vasild** [comment]
> `ENABLE_HARDENING` and the `hardening_interface` have been removed.

Thanks for the note. There was a silent merge conflict.

`a3a799c77c...e83494d75f`: rebase

**Comment by vasild** [comment]
`e83494d75f...63a14e293a`: rebase due to conflicts

**Comment by fanquake** [comment]
> Not sure how many devs would benefit from this. I can trivially set this myself, and I presume the majority isn't using libc++, except for macOS devs? So it may be better to ask macOS devs for review/testing?

What is the status of this. Did this happen?

**Comment by sedited** [comment]
This has not received review in over a year. For a change as small as this one, that gives me the impression that this is not a compelling change for most developers. Maybe an alternative would be adding a small note about setting `LIBCPP_HARDENING_MODE` in the debugging section of the dev notes? I'd imagine that would pass through review quickly.

**Comment by sedited** [comment]
@vasild how do you want to proceed here given my previous [comment](https://github.com/bitcoin/bitcoin/pull/31424#issuecomment-4629552176)?

**Comment by maflcko** [comment]
This has been re-written several times over the last two years, and it is hard to follow the thread, because the comments are about different topics and different approaches.

At this point, it may be best to close this for now. It is always possible to push an alternative to a separate pull or open the changes here in a new clean pull, if reviewers want it.

**Comment by sedited** [comment]
Closing this for now.
